src/copas.lua

```diff
@@ -225,15 +281,18 @@ function copas.receive(client, pattern, part)
   local current_log = _reading_log
   repeat
     s, err, part = client:receive(pattern, part)
-    if s or (not _isTimeout[err]) then
+    if s or (not _isTimeout[err]) or _sockettimeouts.beenhit(client) then
+      _sockettimeouts.reset(client)
       current_log[client] = nil
       return s, err, part
     end
-    if err == "wantwrite" then
+    if err == "wantwrite" then -- Can receive really get a want write?
```

Comment:
yes, upon the sslhandshake, the "wantread" and "wantwrite" errors are specific to LuaSec and are the non-blocking flags indicating what it needs in the process. The sslhandshake function simply adds the sockets to the read/write queues, and hence they end up in these functions here during the handshake.

Reply:
see https://github.com/brunoos/luasec/wiki/LuaSec-0.9#connreceivepattern-prefix

Reply:
This actually is a bigger underlying problem. Copas allows to read and write on the same socket simultaneously (it can be in the reading-set as well as in the writing-set of a call to `select()`).

But this doesn't play nice with ssl, where a renegotiation can force (for example) a reading socket to be pushed into the writing queue, overwriting the socket that was already there, because of the simultaneous writing. I suspect this is the underlying problem for #60 .

Solution seems to be to either:

- not allow renegotiations
- not allow reading/writing simultaneously on an ssl connection

This whole PR turns out to be a lot harder than expected... 🙁